packages/central-server/src/apiV2/surveyResponses/SubmitSurveyResponses.js

```diff
@@ -0,0 +1,59 @@
+/**
+ * Tupaia MediTrak
+ * Copyright (c) 2019 Beyond Essential Systems Pty Ltd
+ */
+
+import { AnalyticsRefresher } from '@tupaia/database';
+import { respond } from '@tupaia/utils';
+import { assertCanSubmitSurveyResponses } from '../import/importSurveyResponses/assertCanImportSurveyResponses';
+import { assertAnyPermissions, assertBESAdminAccess } from '../../permissions';
+import { saveResponsesToDatabase } from '.';
+import { upsertEntitiesAndOptions } from './upsertEntitiesAndOptions';
+import { validateSurveyResponses } from './validateSurveyResponses';
+import { RouteHandler } from '../RouteHandler';
+
+export class SubmitSurveyResponses extends RouteHandler {
+  constructor(req, res) {
+    super(req, res);
+    const { body } = req;
+    this.surveyResponses = Array.isArray(body) ? body : [body];
+  }
+
+  async assertUserHasAccess(transactingModels) {
+    // Check permissions
+    const surveyResponsePermissionsChecker = async accessPolicy => {
+      await assertCanSubmitSurveyResponses(accessPolicy, transactingModels, this.surveyResponses);
+    };
+    await this.assertPermissions(
+      assertAnyPermissions([assertBESAdminAccess, surveyResponsePermissionsChecker]),
+    );
+  }
+
+  async handle() {
+    const { userId } = this.req;
+    const { submitAsPublic, waitForAnalyticsRebuild } = this.query;
+
+    let submitterId = userId;
+    if (submitAsPublic) {
+      const user = await this.models.user.findPublicUser();
+      submitterId = user.id;
+    }
+
+    let results = [];
+
+    await this.models.wrapInTransaction(async transactingModels => {
+      // Upsert entities and options that were created in user's local database
+      await upsertEntitiesAndOptions(transactingModels, this.surveyResponses);
+      await validateSurveyResponses(transactingModels, this.surveyResponses);
+      await this.assertUserHasAccess(transactingModels);
+      results = await saveResponsesToDatabase(transactingModels, submitterId, this.surveyResponses);
```

Comment:
I'm not married to this idea, so feel free to ignore, but since we do something similar in `SurveyResponseImporter`, this could be turned into a reusable function?

Reply:
I had a similar thought, but I think I might leave it as is. The permissions check between the two is slightly different, and I think it will be different again with the ResubmitSurveyResponses route. Since it's just 3 functions and they are neatly independent of each other, I feel this is a good time to let things be a little wetter for the sake of cleaner composition